### DISCUSSION TITLE
How to fix/avoid many tag parsing errors in variadic macros?

### DISCUSSION THREAD

**Discussion by BartVandewoestyne** [microsoft/vscode-cpptools#12645] [Q&A]
I have Debug logging enabled for the C/C++ extension (v1.21.6 in VSCode 1.92.2).  When I open a certain folder/workspace and go check the logs, I see lots of tag parsing errors on Linux kernel code, like for example the following:
```
tag parsing file: /home/bvandewoestyne/projects/urt/Kernel/linux/3.10.31-ltsi/include/uapi/linux/netfilter_arp/arp_tables.h
  tag parsing error (this can be ignored unless symbols can't be found):
    line 30 (col. 51): error: expected a ')'
```
For this specific example, line 30 column 51 is part of a variadic macro:
```
#define ARPT_ENTRY_ITERATE(entries, size, fn, args...) \
        XT_ENTRY_ITERATE(struct arpt_entry, entries, size, fn, ## args)
```
I have the impression that most tag parsing errors occur in such variadic macros.  How can I avoid or fix these tag parsing errors?

**Comment by BartVandewoestyne** [comment]
Seems related to https://github.com/microsoft/vscode-cpptools/issues/9626...

**Reply by sean-mcmanus** [reply]
@BartVandewoestyne Yeah, I believe you can ignore the error message. There are many known cases where it's not really an error (maybe we could change the wording to "warning" so it's less confusing).

**Reply by BartVandewoestyne** [reply]
But isn't it a bug that we get 'tag parsing error' for such variadic macros?  These are supported since C++11 (see https://en.wikipedia.org/wiki/Variadic_macro_in_the_C_preprocessor).

**Reply by sean-mcmanus** [reply]
@BartVandewoestyne Which symbol is not appearing in the Outline view that should be?

**Reply by BartVandewoestyne** [reply]
You are correct, @sean-mcmanus , for the example I gave, the `ARPT_ENTRY_ITERATE` *does* appear in the Outline view.  But then why is an error given when parsing this code???  It seems everything went fine, and still we get an error... Isn't that confusing?

**Reply by sean-mcmanus** [reply]
@BartVandewoestyne Yeah, we added text "this can be ignored unless symbols can't be found" to explain that it may not be an actual error to the user. The layer of code that logs that thinks it's an error, but to a higher level layer it may not be an actual user-facing error. Multiple other users have been confused by that in the past as well. We could potentially change that text to say warning, but then it might hide cases in which there really is an error (that causes symbols to be missing that shouldn't be).

**Reply by BartVandewoestyne** [reply]
Do you know of a way to fix the code and avoid these 'false positive' tag parsing errors (and so only log them if they are *real* errors)?  And so is it just a matter of time before someone implements it?  Or is this a problem for which no one has found a decent solution yet?

**Reply by sean-mcmanus** [reply]
@BartVandewoestyne That logging/error only appears with C_Cpp.loggingLevel is changed from the default of "Error" to "Warning" or higher, so we only intend the logging to be used for diagnostics purposes if an actual bug is experienced (with missing symbols). I don't know of a way to make it differentiate between real errors an non-real errors.
